src/Microsoft.AspNet.SignalR.Client/HeartBeatMonitor.cs

```diff
@@ -129,6 +137,16 @@ private void CheckKeepAlive(TimeSpan timeElapsed)
                     }
                 }
             }
+
+            if (lostConnection)
+            {
+                _connection.Transport.LostConnection(_connection);
```

Comment:
It's probably best to copy the _connection.Transport reference inside the lock, because if _connection.Disconnect() gets invoked between after the lock is released and here, it could null ref.

There's also the possibility that after the lock is released both _connection.Stop() and _connection.Start() could be called. In that case it's better to reference the old Transport would should save the newly-manually-restarted connection from unnecessarily reconnected again thanks to the old transport's cancelled disconnectToken. [At least sometimes.](https://github.com/SignalR/SignalR/blob/44b0f6c5de9cb3afbe2a95b02c4521a9bcabfc40/src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs#L79-L81) I don't miss this code at all.